v1/charm-dev.yaml

```diff
@@ -0,0 +1,76 @@
+version: latest
+
+instances:
+  charm-dev:
+    image: 20.04
+    limits:
+      min-cpu: 2
+      min-mem: 4G
+      min-disk: 10G
+    timeout: 900
+    cloud-init:
+      vendor-data: |
+        snap:
+          commands:
+          - snap install --classic juju
+          - snap install --classic microk8s
+          - snap alias microk8s.kubectl kubectl
+          - snap install --classic charmcraft
+          - snap install yq
+
+        runcmd:
+        - |
+          # disable swap
+          sysctl vm.swappiness=0
+          echo "vm.swappiness = 0" | tee -a /etc/sysctl.conf
+
+        - |
+          # disable unnecessary services
+          systemctl disable man-db.timer man-db.service --now
+          systemctl disable apport.service apport-autoreport.service  --now
+          systemctl disable apt-daily.service apt-daily.timer --now
+          systemctl disable apt-daily-upgrade.service apt-daily-upgrade.timer --now
+          systemctl disable unattended-upgrades.service --now
+          systemctl disable motd-news.service motd-news.timer --now
+          systemctl disable bluetooth.target --now
+          systemctl disable ua-messaging.service ua-messaging.timer --now
+
+        - |
+          # various tools via apt
+          # gh cli
+          curl -fsSL https://cli.github.com/packages/githubcli-archive-keyring.gpg | sudo gpg --dearmor -o /usr/share/keyrings/githubcli-archive-keyring.gpg
+          echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/githubcli-archive-keyring.gpg] https://cli.github.com/packages stable main" | sudo tee /etc/apt/sources.list.d/github-cli.list > /dev/null
+
+          apt-get -y update
+          apt-get -y install jq gh sysstat zsh fzf
+
+        - |
+          # apt cleanup
+          apt-get autoremove -y
+
+        - |
+          # oh-my-zsh + juju plugin
+          homedir=$( getent passwd "ubuntu" | cut -d: -f6 )
+          sudo -u ubuntu sh -c "$(curl -fsSL https://raw.githubusercontent.com/ohmyzsh/ohmyzsh/master/tools/install.sh)" "" --unattended
+          sudo -u ubuntu git clone https://github.com/zsh-users/zsh-autosuggestions.git $homedir/.oh-my-zsh/custom/plugins/zsh-autosuggestions
+          sudo -u ubuntu git clone https://github.com/zsh-users/zsh-syntax-highlighting.git $homedir/.oh-my-zsh/custom/plugins/zsh-syntax-highlighting
+          sudo -u ubuntu sed -i 's/plugins=(git)/plugins=(fzf git zsh-autosuggestions zsh-syntax-highlighting virtualenv colored-man-pages juju colorize)/g' $homedir/.zshrc
+          sudo -u ubuntu sed -i 's/ZSH_THEME="robbyrussell"/ZSH_THEME="bureau"/g' $homedir/.zshrc
+
+        - |
+          # setup charmcraft
+          lxd init --auto
+
+        - |
+          # setup microk8s and bootstrap
+          adduser ubuntu microk8s
+          sudo -u ubuntu microk8s status --wait-ready
+          sudo -u ubuntu microk8s.enable storage
+          sudo -u ubuntu microk8s.enable dns
```

Comment:
Combining results in errors because dns restarts the api server and then you see errors such as 
```
Enabling default storage class
deployment.apps/hostpath-provisioner created
storageclass.storage.k8s.io/microk8s-hostpath created
error when retrieving current configuration of:
Resource: "/v1, Resource=serviceaccounts", GroupVersionKind: "/v1, Kind=ServiceAccount"
Name: "microk8s-hostpath", Namespace: "kube-system"
from server for: "/root/snap/microk8s/2854/tmp/temp.storage.yaml": Get "https://127.0.0.1:16443/api/v1/namespaces/kube-system/serviceaccounts/microk8s-hostpath": dial tcp 127.0.0.1:16443: connect: connection refused
error when retrieving current configuration of:
Resource: "rbac.authorization.k8s.io/v1, Resource=clusterroles", GroupVersionKind: "rbac.authorization.k8s.io/v1, Kind=ClusterRole"
Name: "microk8s-hostpath", Namespace: ""
from server for: "/root/snap/microk8s/2854/tmp/temp.storage.yaml": Get "https://127.0.0.1:16443/apis/rbac.authorization.k8s.io/v1/clusterroles/microk8s-hostpath": dial tcp 127.0.0.1:16443: connect: connection refused
error when retrieving current configuration of:
Resource: "rbac.authorization.k8s.io/v1, Resource=clusterrolebindings", GroupVersionKind: "rbac.authorization.k8s.io/v1, Kind=ClusterRoleBinding"
Name: "microk8s-hostpath", Namespace: ""
from server for: "/root/snap/microk8s/2854/tmp/temp.storage.yaml": Get "https://127.0.0.1:16443/apis/rbac.authorization.k8s.io/v1/clusterrolebindings/microk8s-hostpath": dial tcp 127.0.0.1:16443: connect: connection refused
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?
error: failed to create serviceaccount: Post "https://127.0.0.1:16443/api/v1/namespaces/default/serviceaccounts?fieldManager=kubectl-create": dial tcp 127.0.0.1:16443: connect: connection refused
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?
```